[#1714] Set TTL for connection event messages.

Alfusainey opened the pull request:
fixes #1714

-  I placed the TTL config parameter in the `ClientConfigProperties` since protocol adapters connect to the messaging network to forward the `connection event` downstream. 
- I  currently set a default TTL value of 5 secs. Users can change the default  value  to any non-negative value by changing the corresponding config property.

@sophokles73 I was thinking of allowing  the  option  to set the value to `-1` if  a connection event message should be  keep in the broker (forever) until  it is consumed. WDYT? maybe this value would make a more sensible default?

Signed-off-by: Alfusainey Jallow <alf.jallow@gmail.com>

sophokles73 commented:
@Alfusainey thanks for the PR. We already have a mechanism for configuring the (default and max) TTL to use for normal events being published by devices. The TTL in that case can be configured at the tenant and/or device level.
I do not see why the connection event TTL should be protocol adapter specific, i.e. there is no need to make this configurable per adapter. Instead, it is the northbound applications of a tenant which determine how long the events should be kept in the queue. I therefore believe that the TTL to use for connection events should be configurable at the tenant level with a reasonable default value. It would probably make sense to use the already existing [max-ttl](https://www.eclipse.org/hono/docs/dev/api/tenant/#resource-limits-configuration-format) for that purpose as well ...

Alfusainey commented:
thank kai 👍 I didnt know such a mechanism already exits in hono. i will update the patch

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/DownstreamSender.java:
```diff
@@ -162,6 +166,7 @@
      * @throws IllegalArgumentException if the content type specifies an unsupported character set.
      */
     Future<ProtonDelivery> send(
+            Long maxTtl,
```
I'd rather not introduce a new parameter here but instead simply use the `MessageSender.send(Message)` method for sending the event ...

Alfusainey replied:
I was thinking about this too, but then the code that sets some of properties of the message (body, content-type, to, ...) will be duplicated in the two places. will fix

sophokles73 commented on service-base/src/main/java/org/eclipse/hono/service/monitoring/AbstractMessageSenderConnectionEventProducer.java:
```diff
@@ -82,23 +82,29 @@ protected AbstractMessageSenderConnectionEventProducer(
+                                }
+
+                                final Long maxTtl = tenantObject.result().getResourceLimits().getMaxTtl();
+                                return sender.send(
```
we can use `MessageHelper.newMessage()` here to create the message and send it using `DownstreamSender.send(Message)` ...

Alfusainey replied:
`MessageHelper.newMessage()` sets the message's `orig_adapter` application property to the name of the adapter forwarding the connection event. however, the name of the adapter is already included in the message's payload. i added a `todo` comment in the source code to get some feedback

Alfusainey replied:
@sophokles73 if you have time, can you take a look at rev. 4913146 ?

sophokles73 commented on service-base/src/main/java/org/eclipse/hono/service/monitoring/ConnectionEventProducer.java:
```diff
@@ -52,6 +53,14 @@
+         * 
+         * @return The tenant client instance. This client has to be initialized and started.
+         */
+        TenantClientFactory getTenantClient();
```
IMHO this should be called *getTenantClientFactory* instead

sophokles73 commented on service-base/src/main/java/org/eclipse/hono/service/monitoring/AbstractMessageSenderConnectionEventProducer.java:
```diff
@@ -82,27 +85,45 @@ protected AbstractMessageSenderConnectionEventProducer(
+                                final JsonObject payload = new JsonObject();
+                                payload.put("cause", cause);
+                                payload.put("remote-id", remoteId);
+                                // TODO: can this be removed in favour of 'orig_dapter'?
```
I don't think so because the Event API defines this as a mandatory property of the message body.

sophokles73 commented:
@Alfusainey can you please add a test case verifying the behavior?

Alfusainey commented:
> @Alfusainey can you please add a test case verifying the behavior?

sure, i will do

sophokles73 commented:
@Alfusainey there's a Checkstyle violation in one of the integration tests which causes the Travis build to fail ...

sophokles73 commented on service-base/src/test/java/org/eclipse/hono/service/AbstractProtocolAdapterBaseTest.java:
```diff
@@ -686,6 +696,44 @@ public void verifyConnectionDurationLimitIsReached(final VertxTestContext ctx) {
+    /**
+     * Verifies that the (default) ConnectionEvent API configured for a protocol adapter
+     * sets the connection event message's TTL header value before forwarding the message
+     * to downstream aoplications.
```
applications

sophokles73 commented on service-base/src/test/java/org/eclipse/hono/service/AbstractProtocolAdapterBaseTest.java:
```diff
@@ -686,6 +696,44 @@ public void verifyConnectionDurationLimitIsReached(final VertxTestContext ctx) {
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void tesForwardedConnectionEventMessageHasTtlHeaderSet(final VertxTestContext ctx) {
```
testForwarded ...

sophokles73 commented on service-base/src/test/java/org/eclipse/hono/service/AbstractProtocolAdapterBaseTest.java:
```diff
@@ -686,6 +696,44 @@ public void verifyConnectionDurationLimitIsReached(final VertxTestContext ctx) {
+        final TenantClient tenantClient = mock(TenantClient.class);
+        when(tenantClientFactory.getOrCreateTenantClient()).thenReturn(Future.succeededFuture(tenantClient));
+        final TenantObject tenantObject = TenantObject.from(Constants.DEFAULT_TENANT, true);
+        final ResourceLimits tenantLimits = mock(ResourceLimits.class);
```
there is no need to mock this class, is there?

Alfusainey replied:
no, not really. will be fixed

Alfusainey commented:
> @Alfusainey there's a Checkstyle violation in one of the integration tests which causes the Travis build to fail ... this is not due to your PR and should be resolved once you rebase ...

the unused import  travis is complaining about in the `AmqpUploadTestBase` class is actually used in the class. let me rebase  with master and see

Alfusainey commented:
> @Alfusainey there's a Checkstyle violation in one of the integration tests which causes the Travis build to fail ... this is not due to your PR and should be resolved once you rebase ...

I fixed the changes and rebased this pr with master (since I do not see the unused import travis complaining about). hopefully this will fix the travis build
